02-esc/challenges.js

```diff
@@ -1,143 +1,294 @@
+		challenge_item.appendChild(setParticipants);
+
+		let setDescription = document.createElement("p");
+		setDescription.classList.add("challenge-description");
+		setDescription.textContent = description;
+		challenge_item.appendChild(setDescription);
+
+		let btnBook = document.createElement("button");
+		btnBook.classList.add("btnBook");
+
+		btnBook.setAttribute("data-id", id);
+
+		if (type == "onsite") {
+			btnBook.textContent = "Take challenge online";
+		} else {
+			btnBook.textContent = "Book this room";
+		}
+		challenge_item.appendChild(btnBook);
+
+		btnBook.addEventListener("click", function () {
+			// referens till bokningsfunktion
+		});
+
+		challenge_list.appendChild(challenge_item);
+	}
+}
+
+function doNotReload(e) {
+	// do not want the form to reload the browser window if submitted
+	e.preventDefault();
+}
+
+function clearFilter() {
+	// this function should restore all filters and use the original array to display the rooms
+	filterForm.reset();
+	showAllChallenges(resultFromAPI);
+}
+
+function filterRating(entry, formData) {
+	if (
+		(formData.hasOwnProperty("rating:min") &&
+			+entry.rating < +formData["rating:min"]) ||
+		(formData.hasOwnProperty("rating:max") &&
+			+entry.rating > +formData["rating:max"])
+	) {
+		return false;
+	} else return true;
+}
+
+function filterLabels(entry, key) {
+	if (
+		key.includes("tags:") &&
+		!entry.labels.includes(key.substring(5))
+		// only check the "javascript" part of "tags:javascript", for example
+	) {
+		return false;
+	} else {
+		return true;
+	}
+}
+
+function filterType(entry, formData) {
+	if (
+		formData.hasOwnProperty("type:onsite") &&
+		formData.hasOwnProperty("type:online")
+	) {
+		return true;
+	} else if (formData.hasOwnProperty("type:online") && entry.type != "online") {
+		return false;
+	} else if (formData.hasOwnProperty("type:onsite") && entry.type != "onsite") {
+		return false;
+	} else return true;
+}
+
+function filterSearch(entry, value) {
+	const titleAndDescription = entry.title
+		.concat(" ", entry.description)
+		.toLowerCase()
+		.split(/[ ,]+/);
+	const valueArray = value.toLowerCase().split(/[ ,]+/);
+
+	const trueOrFalse = valueArray.every((value) => {
+		const match = titleAndDescription.find((element) => {
+			if (element.includes(value)) {
+				//Check if the element contains the substring
+				return true;
+			}
+		});
+		return match;
+	});
+
+	if (trueOrFalse) {
+		return true;
+	} else {
+		return false;
+	}
+}
+
+function getFormData() {
+	// get the data from the inputs and convert to object to make it easier to work with
+	const tempFormData = new FormData(filterForm);
+	let formData = {};
+	for (const [key, value] of tempFormData) {
+		if (value !== "") {
+			formData[key] = value;
+		}
+	}
+	return formData;
+}
+
+function useAllFilters(e) {
+	let filteredData = []; // reset array so it can be repopulated after the filter is applied
+
+	const formData = getFormData();
+
+	resultFromAPI.forEach((entry) => {
```

Comment:
Ideally, this `resultFromAPI` variable should be passed to `useAllFilters`. Having the result in a global variable is not great, because it means it could accidentally be changed somewhere, introducing a bug that would be very hard to find because it could be anywhere.